runners/java-fn-execution/src/main/java/org/apache/beam/runners/fnexecution/jobsubmission/JobServerDriver.java

```diff
@@ -0,0 +1,198 @@
+  }
+
+  @Override
+  public void run() {
+    try {
+      jobServer = createJobServer();
+      jobServer.getServer().awaitTermination();
+    } catch (InterruptedException e) {
+      LOG.warn("Job server interrupted", e);
+    } catch (Exception e) {
+      LOG.warn("Exception during job server creation", e);
+    } finally {
+      stop();
+    }
+  }
+
+  // This method is executed by TestPortableRunner via Reflection
+  // Needs to be synchronized to prevent concurrency issues in testing shutdown
+  @SuppressWarnings("WeakerAccess")
+  public synchronized void stop() {
+    if (jobServer != null) {
+      try {
+        jobServer.close();
+        LOG.info("JobServer stopped on {}", jobServer.getApiServiceDescriptor().getUrl());
+        jobServer = null;
+      } catch (Exception e) {
+        LOG.error("Error while closing the jobServer.", e);
+      }
+    }
+    if (artifactStagingServer != null) {
+      try {
+        artifactStagingServer.close();
+        LOG.info(
+            "ArtifactStagingServer stopped on {}",
+            artifactStagingServer.getApiServiceDescriptor().getUrl());
+        artifactStagingServer = null;
+      } catch (Exception e) {
+        LOG.error("Error while closing the artifactStagingServer.", e);
+      }
+    }
+  }
+
+  private GrpcFnServer<InMemoryJobService> createJobServer() throws IOException {
+    InMemoryJobService service = createJobService();
+    GrpcFnServer<InMemoryJobService> jobServiceGrpcFnServer;
+    if (configuration.port == 0) {
+      jobServiceGrpcFnServer = GrpcFnServer.allocatePortAndCreateFor(service, jobServerFactory);
+    } else {
+      Endpoints.ApiServiceDescriptor descriptor =
+          Endpoints.ApiServiceDescriptor.newBuilder()
+              .setUrl(configuration.host + ":" + configuration.port)
+              .build();
+      jobServiceGrpcFnServer = GrpcFnServer.create(service, descriptor, jobServerFactory);
+    }
+    LOG.info("JobService started on {}", jobServiceGrpcFnServer.getApiServiceDescriptor().getUrl());
+    return jobServiceGrpcFnServer;
+  }
+
+  private InMemoryJobService createJobService() throws IOException {
+    artifactStagingServer = createArtifactStagingService();
+    JobInvoker invoker = createJobInvoker();
+    return InMemoryJobService.create(
+        artifactStagingServer.getApiServiceDescriptor(),
+        (String session) -> {
+          try {
+            return BeamFileSystemArtifactStagingService.generateStagingSessionToken(
+                session, configuration.artifactStagingPath);
+          } catch (Exception exn) {
+            throw new RuntimeException(exn);
+          }
+        },
+        (String stagingSessionToken) -> {
+          if (configuration.cleanArtifactsPerJob) {
+            artifactStagingServer.getService().removeArtifacts(stagingSessionToken);
+          }
+        },
+        invoker);
+  }
+
+  private GrpcFnServer<BeamFileSystemArtifactStagingService> createArtifactStagingService()
```

Comment:
would it be possible to have the option to opt out of this artifact staging service? In Samza today we are not leveraging this service and is currently passing in null

Reply:
I suppose we could.

Reply:
I added the option.

Reply:
Great. Thank you!!

Reply:
I think it is ok to not enable artifact staging temporarily. It is up to the Runner author to decide that. I don't think it should be exposed as a general option.

Reply:
Couldn't we handle this internally by the JobServerDriver's implementation?

Reply:
I think it makes sense to expose this as a general option because many runners will want to do that (and in the same way, via a BeamArtifactsServer), but allow it to be disabled. The goal of this library is to share as much common code as possible, and artifact serving will be common (though possibly not ubiquitous).

Reply:
@mxm  if there is no option to disable the artifact staging, then basically Samza runner would need to reimplement most of the codes in this file. I see this refactor as a good opportunity to remove much of the duplicate codes between Flink and Samza, and potentially future portable runners.     I agree it isn't super nice to expose this as a general option to users. Alternatively we could abstract this out as a protect method so that we could override the behavior. Either way is fine by me

Reply:
>The goal of this library is to share as much common code as possible, and artifact serving will be common (though possibly not ubiquitous).

I'm all for sharing as much code as possible but I don't think we should sacrifice the integrity of our execution model for it. Artifact staging is a fixed part of portability and should not be skippable by users.

>Alternatively we could abstract this out as a protect method so that we could override the behavior.

I think that is the way to go. We shouldn't leak the internal design choices of the Runners into a general interface.

Reply:
Alright, I made `createJobService` protected so runners can optionally override it.

Reply:
The Samza runner can do something like this:

```java
  @Override
  protected InMemoryJobService createJobService() throws IOException {
    return InMemoryJobService.create(
        null,
        this::createSessionToken,
        stagingSessionToken -> {},
        createJobInvoker());
  }
```

Reply:
Looks good. Thanks!